File: src/engine/packet.ts
/**
 * Frame packet produced by the Rust engine (see crates/overtone-core/src/packet.rs).
 * All values little-endian.
 */
export const PACKET_MAGIC = 0x4e54564f;
export const PACKET_VERSION = 2;
export const HEADER_LEN = 68;
export const FLAG_NEW_AUDIO = 1;
export const FLAG_LIVE = 2;
export const FLAG_FULL = 4;

export interface FramePacket {
  seq: number;
  height: number;
  flags: number;
  pitchHz: number;
  targetHz: number;
  /** Index into EngineInfo.notes, or -1. */
  note: number;
  sampleRate: number;
  /** Pixels to move the existing timeline right (negative = left). */
  shift: number;
  /** x position of the first included column. */
  columnStart: number;
  /** Number of timeline columns included. */
  columns: number;
  /** Timeline width the engine rendered for. */
  width: number;
  /** Time at the newest edge of the view, seconds. */
  viewEnd: number;
  historyStart: number;
  historyEnd: number;
  pxPerSecond: number;
  /** RGBA, row-major, `columns × height`. */
  pixels: Uint8ClampedArray;
  /** Live bar length per row, index 0 = top row. */
  live: Float32Array;
}

export function parsePacket(bytes: Uint8Array): FramePacket {
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  if (view.getUint32(0, true) !== PACKET_MAGIC) {
    throw new Error("bad frame packet magic");
  }
  const version = view.getUint32(4, true);
  if (version !== PACKET_VERSION) {
    throw new Error(`unsupported frame packet version ${version}`);
  }
  const height = view.getUint32(12, true);
  const columns = view.getUint32(40, true);
  const pixelsOffset = bytes.byteOffset + HEADER_LEN;
  const pixelsLen = columns * height * 4;
  const liveOffset = pixelsOffset + pixelsLen;
  // Float32Array needs 4-byte alignment; copy when the packet is not aligned.
  const live =
    liveOffset % 4 === 0
      ? new Float32Array(bytes.buffer, liveOffset, height)
      : new Float32Array(bytes.buffer.slice(liveOffset, liveOffset + height * 4));
  return {
    seq: view.getUint32(8, true),
    height,
    flags: view.getUint32(16, true),
    pitchHz: view.getFloat32(20, true),
    targetHz: view.getFloat32(24, true),
    note: view.getInt32(28, true),
    sampleRate: view.getFloat32(32, true),
    shift: view.getInt32(36, true),
    columnStart: view.getUint32(64, true),
    columns,
    width: view.getUint32(44, true),
    viewEnd: view.getFloat32(48, true),
    historyStart: view.getFloat32(52, true),
    historyEnd: view.getFloat32(56, true),
    pxPerSecond: view.getFloat32(60, true),
    pixels: new Uint8ClampedArray(bytes.buffer, pixelsOffset, pixelsLen),
    live,
  };
}

File: crates/overtone-core/src/packet.rs
//! Binary layout of the per-frame packet handed to the frontend.
//!
//! All integers are little-endian. Layout (byte offsets):
//!
//! | offset | type | field |
//! |--------|------|-------|
//! | 0  | u32 | magic `0x4E54564F` ("OVTN") |
//! | 4  | u32 | version (2) |
//! | 8  | u32 | frame sequence number |
//! | 12 | u32 | height (rows) |
//! | 16 | u32 | flags (bit 0: new audio, bit 1: following live, bit 2: full redraw) |
//! | 20 | f32 | detected pitch (Hz, 0 = none) |
//! | 24 | f32 | nearest note frequency (Hz, 0 = none) |
//! | 28 | i32 | nearest note index into `info.notes` (-1 = none) |
//! | 32 | f32 | sample rate |
//! | 36 | i32 | shift: pixels to move the existing timeline right (negative = left) |
//! | 40 | u32 | columns: number of timeline columns included (== width for a full redraw) |
//! | 44 | u32 | width of the timeline in pixels |
//! | 64 | u32 | column_start: x position of the first included column |
//! | 48 | f32 | time at the left (newest) edge of the view, seconds |
//! | 52 | f32 | oldest recorded time, seconds |
//! | 56 | f32 | newest recorded time, seconds |
//! | 60 | f32 | timeline scale, pixels per second |
//! | 68 | u8 × 4·columns·height | timeline pixels, RGBA, row-major, x = 0 is newest |
//! | 68 + 4·columns·height | f32 × height | live bar length per row, index 0 = top row |

pub const MAGIC: u32 = 0x4E54_564F;
pub const VERSION: u32 = 2;
pub const HEADER_LEN: usize = 68;
pub const FLAG_NEW_AUDIO: u32 = 1;
pub const FLAG_LIVE: u32 = 2;
pub const FLAG_FULL: u32 = 4;

pub const fn columns_offset() -> usize {
    HEADER_LEN
}

pub const fn live_offset(columns: usize, height: usize) -> usize {
    HEADER_LEN + columns * height * 4
}

/// Length of a packet carrying `columns` timeline columns.
pub const fn packet_len(columns: usize, height: usize) -> usize {
    live_offset(columns, height) + height * 4
}

pub struct Header {
    pub seq: u32,
    pub height: u32,
    pub flags: u32,
    pub pitch_hz: f32,
    pub target_hz: f32,
    pub note: i32,
    pub sample_rate: f32,
    pub shift: i32,
    pub column_start: u32,
    pub columns: u32,
    pub width: u32,
    pub view_end: f32,
    pub history_start: f32,
    pub history_end: f32,
    pub px_per_second: f32,
}

impl Header {
    pub fn write(&self, out: &mut [u8]) {
        debug_assert!(out.len() >= HEADER_LEN);
        out[0..4].copy_from_slice(&MAGIC.to_le_bytes());
        out[4..8].copy_from_slice(&VERSION.to_le_bytes());
        out[8..12].copy_from_slice(&self.seq.to_le_bytes());
        out[12..16].copy_from_slice(&self.height.to_le_bytes());
        out[16..20].copy_from_slice(&self.flags.to_le_bytes());
        out[20..24].copy_from_slice(&self.pitch_hz.to_le_bytes());
        out[24..28].copy_from_slice(&self.target_hz.to_le_bytes());
        out[28..32].copy_from_slice(&self.note.to_le_bytes());
        out[32..36].copy_from_slice(&self.sample_rate.to_le_bytes());
        out[36..40].copy_from_slice(&self.shift.to_le_bytes());
        out[40..44].copy_from_slice(&self.columns.to_le_bytes());
        out[44..48].copy_from_slice(&self.width.to_le_bytes());
        out[48..52].copy_from_slice(&self.view_end.to_le_bytes());
        out[52..56].copy_from_slice(&self.history_start.to_le_bytes());
        out[56..60].copy_from_slice(&self.history_end.to_le_bytes());
        out[60..64].copy_from_slice(&self.px_per_second.to_le_bytes());
        out[64..68].copy_from_slice(&self.column_start.to_le_bytes());
    }
}
